File: DesignPatternsExercises/ruby/mediator_design_pattern/solution/mediator.rb
require_relative 'currency_converter'

class Mediator
  def initialize
    @buyer_list = []
    @seller_list = []
  end

  def add_seller(seller)
    @seller_list << seller
  end

  def add_buyer(buyer)
    @buyer_list << buyer
  end

  # def __________(5)__________(currency, bid)
  def place_bid(currency, bid)
    sellers = []
    @seller_list.each { |seller| sellers << seller if seller.currency == currency && seller.selling_price > 0 }

    if sellers.empty?
      puts "Seller is not found for #{bid} #{currency}"
    else
      # dollar_amount = CurrencyConverter.__________(6)__________(currency, bid)
      dollar_amount = CurrencyConverter.convert(currency, bid)
      sellers.each { |seller| return if seller.bid_accepted?(dollar_amount) }
    end
  end
end

File: DesignPatternsExercises/java/MediatorDesignPattern/Exercise/Mediator.java
import java.util.ArrayList;

public class Mediator {
	private ArrayList<Buyer> buyersList = new ArrayList<>();
	private ArrayList<Seller> sellersList = new ArrayList<>();

	public void addSeller(Seller seller) {
		sellersList.add(seller);
	}

	public void addBuyer(Buyer buyer) {
		buyersList.add(buyer);
	}

	// public void __________(4)__________(String currency, float bid) {
	public void placeBid(String currency, float bid) {
		ArrayList<Seller> sellers = new ArrayList<>();
		for (Seller seller : sellersList) {
			if (seller.getCurrency().equals(currency) && seller.getSellingPrice() > 0) {
				sellers.add(seller);
			}
		}
		if (sellers.isEmpty()) {
			System.out.println("Seller is not found for " + bid + " " + currency);
		} else {
			// float dollarAmount = CurrencyConverter.__________(5)__________(currency, bid);
			float dollarAmount = CurrencyConverter.convert(currency, bid);
			for (Seller seller : sellers) {
				if (seller.isBidAccepted(dollarAmount))
					return;
			}
		}
	}
}
